lib/dynamoid/transaction/upsert.rb

```diff
@@ -0,0 +1,33 @@
+require_relative 'action'
+
+module Dynamoid
+  module Transaction
+    class Upsert < Action
+
+      def run_callbacks
+        (yield; return) unless model
+        model.run_callbacks(:save) do
+          model.run_callbacks(model.new_record? ? :create : :update) do # guessing if it's new or not! (could already be in db)
+            model.run_callbacks(:validate) do
+              yield
+            end
+          end
+        end
+      end
+
+      def to_h
+        item = attributes.clone
+        item = add_timestamps(item)
+        item = model.attributes.merge(item) if model
+        item[model_class.hash_key] ||= SecureRandom.uuid
+        item = Dynamoid::Dumping.dump_attributes(item, model_class.attributes)
+        {
+          put: {
+            item: sanitize_item(item),
+            table_name: model_class.table_name,
+          }
+        }
+      end
```

Comment:
Class-level methods like `create`/`update`/`upsert` accepts keys and attributes as separate parameters:

```ruby
 User.upsert('1', age: 26)
```

This way we can find the proper item in DynamoDB even when user updates sort key for instance (can he change a partition key? probably he can)

Reply:
Yes! I changed update and delete to take an id which can be a hash_key or an [hash_key, range_key] like is used by find(). I don't think id is needed for create and upsert since those are both putting complete records. Should we include id anyways for those?

Reply:
@andrykonchin I assumed that the partition key could be changed but that is not true. A record must be deleted and recreated if you wish to change the partition key. How would you feel about simplifying the APIs for update and upsert?

Instead of:
`txn.update!(User, '1', {name: 'bob'} )`
it would be:
`txn.update!(User, {id: '1', name: 'bob'} )`

When there's a range key:
`txn.update!(User, ['1', 26], {name: 'bob'} )`
would be:
`txn.update!(User, {id: '1', age: 26, name: 'bob'} )`

This makes the API similar to create which I like but perhaps it's too inconsistent with the other update and upserts.

Reply:
> This makes the API similar to create which I like but perhaps it's too inconsistent with the other update and upserts.

It seems to me it will not work for updating a range key attribute. Its current value should be used to find an item and a new value to update.

Reply:
No part of a primary key can be updated. The record must be deleted and re-created to change the key. If it's a composite key this includes the range key which cannot be changed.

I see your point about wanting to be consistent with the non-transaction methods however not allowing a separate key to be specified may be a better direction to go in. This will discourage people from trying to update primary keys among other things. I like the shorter method signature and that all the transaction method signatures are more similar as well.

Reply:
> No part of a primary key can be updated. The record must be deleted and re-created to change the key. If it's a composite key this includes the range key which cannot be changed.

Indeed, exception is raised:
```
One or more parameter values were invalid: Cannot update attribute age. This attribute is part
 of the key (Aws::DynamoDB::Errors::ValidationException)
```